test/cmdlineTests/storage_layout_transient_value_types/output.json

```diff
@@ -0,0 +1,19 @@
+{
+    "errors": [
+        {
+            "component": "general",
+            "formattedMessage": "Internal exception in StandardCompiler::compile: /solidity/libsolidity/interface/StorageLayout.cpp(49): Throw in function solidity::Json solidity::frontend::StorageLayout::generate(const solidity::frontend::VariableDeclaration&, const u256&, unsigned int)
+Dynamic exception type: boost::wrapexcept<solidity::langutil::UnimplementedFeatureError>
+std::exception::what: Unimplemented feature
+[solidity::util::tag_comment*] = Unimplemented feature
+",
```

Comment:
I decide to just remove it. `StorageLayout` does not have a `errorReporter` member, so I think it is simpler to just not test this for now.
However, this made me question about other outputs from codegen which are being generated basically ignoring `transient` and assuming `storage`. Should we also throw `UnimplementedFeatureError` for those cases too?

Reply:
I was thinking that we could also throw a `fatalTypeError` during type checking, I guess. Or just throw  `UnimplementedFeatureError` in type checking but then most of the tests will have to be removed (and saved for later) and we could add tests using the `stopAfter`  flag set to `parsing`...

Reply:
Depends on the output. If there are any others that are wrong with transient storage, they should fail. But I don't think we need to do that with the whole type checking. The variables are getting `transient` location in the AST so it's clear what's happening. 

It's the cases like storage layout that are bad - the compiler was giving outright wrong information - telling you that transient variables live in storage, and using wrong slots for the actual storage variables.

Reply:
I was thinking about codegen output, like, for example, opcodes, asm and bin which will just generate code considering the variables to located on storage.
Compiling this with `solc --opcodes`: 
```solidity
contract C {
    uint transient x = 42;
}
```
generates output which uses `SSTORE` for the `x` var:
```
Opcodes:
PUSH1 0x80 PUSH1 0x40 MSTORE PUSH1 0x2A PUSH0 SSTORE ...
```

Reply:
Ah, great point. I somehow assumed that this would already be happening here, but you're totally right, we only added unimplemented asserts for stuff like reference types. Value types are allowed and will proceed to the codegen and just ignore the location. We definitely should block that, because otherwise anyone who mistakenly assumes that transient is fully implemented and starts using it will get a bad surprise. Declaring a variable as transient and getting code using storage variable instead is something we'd normally consider a serious codegen error.

The main complication though is that if we add asserts, syntax tests will start failing because they are hard-coded to run compilation: https://github.com/ethereum/solidity/blob/028cf3309437819503a37f60bfa8aad2af7fb9b7/test/libsolidity/SyntaxTest.cpp#L75 We'd indeed need support for `stopAfter: parsing` in syntax tests.

So the choices here are to add this as a setting you can enable in certain syntax tests and then use it in this PR or to just leave this PR unmerged until the codegen part is implemented. Fortunately it did not make it into 0.8.26, but there's still no 100% certainty we'll have the rest ready by 0.8.27 so merging it otherwise is risky.

Reply:
Yeah, should be quite simple. It's just a matter of parsing the new setting in [`SyntaxTest` constructor](https://github.com/ethereum/solidity/blob/v0.8.25/test/libsolidity/SyntaxTest.cpp#L40-L54) and later passing the selected target stage to `AnalysisFramework` in `SyntaxTest::parseAndAnalyze()`.

I'd recommend doing that in a separate PR though.

And I think it'll generally be useful for future cases where we'll want to implement only parsing or only analysis. It would already have been useful for experimental - we had the same problem back then and we solved it by hacking in a condition for experimental, but this would have been a more general solution.

Reply:
I added the cli test back since we are catching `unimplementedFeatureError`s thrown by the compiler and reporting them as errors (albeit for now there's no source location in those).